src/audio/mixer/mixer.c

```diff
@@ -119,33 +121,39 @@ static int mixer_process(struct processing_module *mod,
 		if (!audio_stream_set_zero(mod->output_buffers[0].data, sink_bytes))
 			mod->output_buffers[0].size = sink_bytes;
 
-		md->sources_inactive = true;
-
 		return 0;
 	}
 
 	/* Every source has the same format, so calculate bytes based on the first one */
 	source_bytes = frames * audio_stream_frame_bytes(mod->input_buffers[0].data);
 
-	if (md->sources_inactive) {
-		md->sources_inactive = false;
-		comp_dbg(dev, "mixer_copy exit sources_inactive state");
-	}
-
 	sink_bytes = frames * audio_stream_frame_bytes(mod->output_buffers[0].data);
 
 	comp_dbg(dev, "mixer_process(), source_bytes = 0x%x, sink_bytes = 0x%x",
 		 source_bytes, sink_bytes);
 
 	/* mix streams */
-	for (i = 0; i < num_input_buffers; i++)
-		sources_stream[i] = mod->input_buffers[i].data;
+	int j = 0;
+
+	for (i = 0; i < num_input_buffers; i++) {
+		uint32_t avail_frames;
+
+		avail_frames = audio_stream_avail_frames_aligned(mod->input_buffers[i].data,
+								 mod->output_buffers[0].data);
+
+		/* if one source is inactive, skip it */
+		if (avail_frames == 0)
+			continue;
+
+		sources_stream[j++] = mod->input_buffers[i].data;
+	}
 
-	md->mix_func(dev, mod->output_buffers[0].data, sources_stream, num_input_buffers, frames);
+	if (j)
+		md->mix_func(dev, mod->output_buffers[0].data, sources_stream, j, frames);
```

Comment:
Perhaps I should use active_input_buffers here and below?

Reply:
You need j anyway to have the exact active buffers, so you can continue using j here and below

Reply:
Hmm, wouldn't j==active_input_buffers ? If j can be differnent, then I think it deserves a longer more descriptive name as it's used across this long function (i.e. not just a loop index variable anymore).

Reply:
They can differ if there's a race changing which buffers are active between the first check and this loop. I can write an assert stating that they don't differ I suppose.